File: rust/vineyard/src/client/ipc_client.rs
// Copyright 2020-2023 Alibaba Group Holding Limited.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::HashMap;
use std::io;
use std::net::Shutdown;
use std::os::unix::net::UnixStream;
use std::sync::Arc;
use std::sync::Mutex;

use arrow_buffer::Buffer;
use parking_lot::ReentrantMutex;
use parking_lot::ReentrantMutexGuard;

use crate::common::util::arrow::*;
use crate::common::util::protocol::*;
use crate::common::util::status::*;
use crate::common::util::typename::*;
use crate::common::util::uuid::*;

use super::client::*;
use super::ds::blob::{Blob, BlobWriter};
use super::ds::object::*;
use super::ds::object_meta::ObjectMeta;
use super::io::*;

mod memory {

    use std::collections::{hash_map, HashMap, HashSet};
    use std::fs::File;
    use std::os::fd::{AsRawFd, FromRawFd};
    use std::os::unix::net::UnixStream;

    use memmap2::{Mmap, MmapMut, MmapOptions};

    use crate::common::memory::fling::recv_fd;
    use crate::common::util::status::*;

    #[derive(Debug)]
    pub struct MmapEntry {
        fd: File,
        ro_pointer: *const u8,
        rw_pointer: *mut u8,
        length: usize,

        mmap_readonly: Option<Mmap>,
        mmap_readwrite: Option<MmapMut>,
    }

    impl MmapEntry {
        pub fn new(fd: i32, map_size: usize, realign: bool) -> Self {
            let size = if realign {
                map_size - std::mem::size_of::<usize>()
            } else {
                map_size
            };
            return MmapEntry {
                fd: unsafe { File::from_raw_fd(fd) },
                ro_pointer: std::ptr::null(),
                rw_pointer: std::ptr::null_mut(),
                length: size,
                mmap_readonly: None,
                mmap_readwrite: None,
            };
        }

        #[allow(dead_code)]
        pub fn fd(&self) -> i32 {
            return self.fd.as_raw_fd();
        }

        pub fn map(&mut self) -> Result<*const u8> {
            if self.ro_pointer.is_null() {
                let mmap = unsafe { MmapOptions::new().len(self.length).offset(0).map(&self.fd) }?;
                self.ro_pointer = mmap.as_ptr();
                self.mmap_readonly = Some(mmap);
            }
            return Ok(self.ro_pointer);
        }

        pub fn map_mut(&mut self) -> Result<*mut u8> {
            if self.rw_pointer.is_null() {
                let mut mmap = unsafe { MmapOptions::new().len(self.length).map_mut(&self.fd) }?;
                self.rw_pointer = mmap.as_mut_ptr();
                self.mmap_readwrite = Some(mmap);
            }
            return Ok(self.rw_pointer);
        }
    }

    #[derive(Debug)]
    pub struct MmapManager {
        entries: HashMap<i32, MmapEntry>,
    }

    impl MmapManager {
        pub fn new() -> Self {
            return MmapManager {
                entries: HashMap::new(),
            };
        }

        pub fn mmap(
            &mut self,
            stream: &UnixStream,
            fd: i32,
            map_size: usize,
            realign: bool,
        ) -> Result<*const u8> {
            if let hash_map::Entry::Vacant(entry) = self.entries.entry(fd) {
                entry.insert(MmapEntry::new(recv_fd(stream)?, map_size, realign));
            }
            match self.entries.get_mut(&fd) {
                Some(entry) => {
                    return entry.map();
                }
                None => {
                    return Err(VineyardError::invalid(format!(
                        "Failed to find mmap entry for fd even after insert: {}",
                        fd
                    )));
                }
            }
        }

        pub fn mmap_mut(
            &mut self,
            stream: &UnixStream,
            fd: i32,
            map_size: usize,
            realign: bool,
        ) -> Result<*mut u8> {
            if let hash_map::Entry::Vacant(entry) = self.entries.entry(fd) {
                entry.insert(MmapEntry::new(recv_fd(stream)?, map_size, realign));
            }
            match self.entries.get_mut(&fd) {
                Some(entry) => {
                    return entry.map_mut();
                }
                None => {
                    return Err(VineyardError::invalid(format!(
                        "Failed to find mmap entry for fd even after insert: {}",
                        fd
                    )));
                }
            }
        }

        #[allow(dead_code)]
        pub fn exists(&self, fd: i32) -> i32 {
            if self.entries.contains_key(&fd) {
                return -1;
            } else {
                return fd;
            }
        }

        #[allow(dead_code)]
        pub fn deduplicate(&self, fd: i32, fds: &mut Vec<i32>, dedup: &mut HashSet<i32>) {
            if !dedup.contains(&fd) && !self.entries.contains_key(&fd) {
                fds.push(fd);
                dedup.insert(fd);
            }
        }
    }
}

#[derive(Debug)]
pub struct IPCClient {
    connected: bool,
    pub ipc_socket: String,
    pub rpc_endpoint: String,
    pub instance_id: InstanceID,
    pub server_version: String,
    pub support_rpc_compression: bool,

    stream: UnixStream,
    lock: ReentrantMutex<()>,
    mmap: memory::MmapManager,
}

impl Drop for IPCClient {
    fn drop(&mut self) {
        self.disconnect();
    }
}

unsafe impl Send for IPCClient {}
unsafe impl Sync for IPCClient {}

impl Client for IPCClient {
    fn disconnect(&mut self) {
        if !self.connected() {
            return;
        }
        self.connected = false;
        if let Ok(message_out) = write_exit_request() {
            if let Err(err) = self.do_write(&message_out) {
                error!("Failed to disconnect the client: {}", err);
            }
        }
        self.stream.shutdown(Shutdown::Both).unwrap_or_else(|e| {
            error!("Failed to shutdown IPCClient stream: {}", e);
        });
    }

    #[cfg(not(feature = "nightly"))]
    fn connected(&mut self) -> bool {
        return self.connected;
    }

    #[cfg(feature = "nightly")]
    fn connected(&mut self) -> bool {
        if self.stream.set_nonblocking(true).is_err() {
            return false;
        }
        match self.stream.peek(&mut [0]) {
            Ok(_) => {
                let _ = self.stream.set_nonblocking(false);
                return true;
            }
            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
                let _ = self.stream.set_nonblocking(false);
                return true;
            }
            Err(_) => {
                self.connected = false;
                return false;
            }
        }
    }

    fn ensure_connect(&mut self) -> Result<ReentrantMutexGuard<'_, ()>> {
        if !self.connected() {
            return Err(VineyardError::io_error("client not connected"));
        }
        return Ok(self.lock.lock());
    }

    fn do_read(&mut self) -> Result<String> {
        return do_read(&mut self.stream);
    }

    fn do_write(&mut self, message_out: &str) -> Result<()> {
        return do_write(&mut self.stream, message_out);
    }

    fn instance_id(&self) -> InstanceID {
        return self.instance_id;
    }

    fn create_metadata(&mut self, metadata: &ObjectMeta) -> Result<ObjectMeta> {
        let mut meta = metadata.clone();
        meta.set_instance_id(self.instance_id());
        meta.set_transient(true);
        if !meta.has_key("nbytes") {
            meta.set_nbytes(0usize);
        }
        if meta.is_incomplete() {
            let _ = self.sync_metadata();
        }
        let (id, signature, instance_id) = self.create_data(meta.meta_data())?;
        meta.set_id(id);
        meta.set_signature(signature);
        meta.set_instance_id(instance_id);
        if meta.is_incomplete() {
            meta = self.get_metadata(id)?;
        }
        meta.set_client(self);
        return Ok(meta);
    }

    fn get_metadata(&mut self, id: ObjectID) -> Result<ObjectMeta> {
        let data = self.get_data(id, false, false)?;
        let mut meta = ObjectMeta::new(self, data)?;

        let buffer_id_vec: Vec<ObjectID> =
            meta.get_buffers().buffer_ids().iter().cloned().collect();
        let buffers = self.get_buffers(&buffer_id_vec, false)?;
        for (buffer_id, buffer) in buffers {
            meta.set_buffer(buffer_id, buffer)?;
        }
        return Ok(meta);
    }

    fn get_metadata_batch(&mut self, ids: &[ObjectID]) -> Result<Vec<ObjectMeta>> {
        let data_vec = self.get_data_batch(ids)?;
        let mut metadatas = Vec::new();
        let mut buffer_id_vec: Vec<ObjectID> = Vec::new();
        for data in data_vec {
            let meta = ObjectMeta::new(self, data)?;
            buffer_id_vec.extend(meta.get_buffers().buffer_ids());
            metadatas.push(meta);
        }

        let buffers = self.get_buffers(&buffer_id_vec, false)?;
        for meta in metadatas.iter_mut() {
            for buffer_id in meta.get_buffers().buffer_ids().clone() {
                if let Some(buffer) = buffers.get(&buffer_id) {
                    meta.set_buffer(buffer_id, buffer.clone())?;
                }
            }
        }
        return Ok(metadatas);
    }
}

impl IPCClient {
    #[allow(clippy::should_implement_trait)]
    pub fn default() -> Result<IPCClient> {
        let default_ipc_socket = std::env::var(VINEYARD_IPC_SOCKET_KEY)?;
        return IPCClient::connect(&default_ipc_socket);
    }

    pub fn connect(socket: &str) -> Result<IPCClient> {
        let mut stream = connect_ipc_socket_retry(&socket)?;
        let message_out = write_register_request(RegisterRequest {
            version: VERSION.into(),
            store_type: "Normal".into(),
            session_id: 0,
            username: String::new(),
            password: String::new(),
            support_rpc_compression: false,
        })?;
        do_write(&mut stream, &message_out)?;
        let reply = read_register_reply(&do_read(&mut stream)?)?;
        return Ok(IPCClient {
            connected: true,
            ipc_socket: reply.ipc_socket,
            rpc_endpoint: reply.rpc_endpoint,
            instance_id: reply.instance_id,
            server_version: reply.version,
            support_rpc_compression: reply.support_rpc_compression,
            stream: stream,
            lock: ReentrantMutex::new(()),
            mmap: memory::MmapManager::new(),
        });
    }

    pub fn create_blob(&mut self, size: usize) -> Result<BlobWriter> {
        let (id, buffer) = self.create_buffer(size)?;
        return Ok(BlobWriter::new(id, buffer));
    }

    pub fn get_blob(&mut self, id: ObjectID) -> Result<Blob> {
        let buffer = self.get_buffer(id, false)?;
        let size = match &buffer {
            Some(buffer) => buffer.len(),
            None => 0,
        };
        let mut meta = ObjectMeta::new_from_typename(typename::<Blob>());
        meta.set_id(id);
        meta.set_instance_id(self.instance_id());
        meta.set_or_add_buffer(id, buffer.clone())?;
        return Ok(Blob::new(meta, size, buffer));
    }

    fn create_buffer(&mut self, size: usize) -> Result<(ObjectID, Option<Buffer>)> {
        if size == 0 {
            return Ok((empty_blob_id(), Some(arrow_buffer_null())));
        }
        let _ = self.ensure_connect()?;
        let message_out = write_create_buffer_request(size)?;
        self.do_write(&message_out)?;
        let reply = read_create_buffer_reply(&self.do_read()?)?;
        if reply.payload.data_size == 0 {
            return Ok((reply.id, Some(arrow_buffer_null())));
        }
        let pointer = self.mmap.mmap_mut(
            &self.stream,
            reply.payload.store_fd,
            reply.payload.map_size,
            true,
        )?;
        let buffer =
            arrow_buffer_with_offset(pointer, reply.payload.data_offset, reply.payload.data_size);
        return Ok((reply.id, Some(buffer)));
    }

    fn get_buffer(&mut self, id: ObjectID, unsafe_: bool) -> Result<Option<Buffer>> {
        let buffers = self.get_buffers(&[id], unsafe_)?;
        return buffers
            .get(&id)
            .cloned()
            .ok_or(VineyardError::object_not_exists(format!(
                "buffer {} doesn't exist",
                id
            )));
    }

    fn get_buffers(
        &mut self,
        ids: &[ObjectID],
        unsafe_: bool,
    ) -> Result<HashMap<ObjectID, Option<Buffer>>> {
        let _ = self.ensure_connect()?;
        let message_out = write_get_buffers_request(&ids, unsafe_)?;
        self.do_write(&message_out)?;
        let reply = read_get_buffers_reply(&self.do_read()?)?;

        let mut buffers = HashMap::new();
        for payload in reply.payloads {
            if payload.data_size == 0 {
                buffers.insert(payload.object_id, Some(arrow_buffer_null()));
                continue;
            }
            let pointer = self
                .mmap
                .mmap(&self.stream, payload.store_fd, payload.map_size, true)?;
            let buffer = arrow_buffer_with_offset(pointer, payload.data_offset, payload.data_size);
            buffers.insert(payload.object_id, Some(buffer));
        }
        return Ok(buffers);
    }

    pub fn get<T: Object + Create>(&mut self, id: ObjectID) -> Result<Box<T>> {
        let meta = self.get_metadata(id)?;
        let mut object = T::create();
        object.construct(meta)?;
        return downcast_object(object);
    }

    pub fn fetch_and_get<T: Object + Create>(&mut self, id: ObjectID) -> Result<Box<dyn Object>> {
        let meta = self.fetch_and_get_metadata(id)?;
        let mut object = T::create();
        object.construct(meta)?;
        return Ok(object);
    }
}

pub struct IPCClientManager {}

impl IPCClientManager {
    pub fn get_default() -> Result<Arc<Mutex<IPCClient>>> {
        let default_ipc_socket = std::env::var(VINEYARD_IPC_SOCKET_KEY)?;
        return IPCClientManager::get(default_ipc_socket);
    }

    pub fn get<S: Into<String>>(socket: S) -> Result<Arc<Mutex<IPCClient>>> {
        let mut clients = IPCClientManager::get_clients().lock()?;
        let socket = socket.into();
        if let Some(client) = clients.get(&socket) {
            if client.lock()?.connected() {
                return Ok(client.clone());
            }
        }
        let client = Arc::new(Mutex::new(IPCClient::connect(&socket)?));
        clients.insert(socket, client.clone());
        return Ok(client);
    }

    pub fn close<S: Into<String>>(socket: S) -> Result<()> {
        let mut clients = IPCClientManager::get_clients().lock()?;
        let socket = socket.into();
        if let Some(client) = clients.get(&socket) {
            if Arc::strong_count(client) == 1 {
                clients.remove(&socket);
            }
            return Ok(());
        } else {
            return Err(VineyardError::invalid(format!(
                "Failed to close the client due to the unknown socket: {}",
                socket
            )));
        }
    }

    fn get_clients() -> &'static Arc<Mutex<HashMap<String, Arc<Mutex<IPCClient>>>>> {
        lazy_static! {
            static ref CONNECTED_CLIENTS: Arc<Mutex<HashMap<String, Arc<Mutex<IPCClient>>>>> =
                Arc::new(Mutex::new(HashMap::new()));
        }
        return &CONNECTED_CLIENTS;
    }
}

#[macro_export]
macro_rules! get {
    ($client: ident, $object_ty: ty, $object_id: expr) => {
        $client.get::<$object_ty>($object_id)
    };
}

#[macro_export]
macro_rules! put {
    ($client: expr, $builder_ty: ty, $($arg: expr),* $(,)?) => {
        match <$builder_ty>::new($client, $($arg),*) {
            Ok(builder) => builder.seal($client),
            Err(e) => Err(e),
        }
    };
}

File: go/vineyard/pkg/client/ipc_client.go
/** Copyright 2020-2023 Alibaba Group Holding Limited.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package client

import (
	"net"
	"syscall"

	arrow "github.com/apache/arrow/go/v11/arrow/memory"
	"github.com/pkg/errors"

	"github.com/v6d-io/v6d/go/vineyard/pkg/client/io"
	"github.com/v6d-io/v6d/go/vineyard/pkg/common"
	"github.com/v6d-io/v6d/go/vineyard/pkg/common/memory"
	"github.com/v6d-io/v6d/go/vineyard/pkg/common/types"
)

type IPCClient struct {
	*ClientBase
	unixConn  *net.UnixConn
	mmapTable map[int]MmapEntry
}

type MmapEntry struct {
	FD   int
	Size int64
	ro   []byte // slice
	rw   []byte // slice
}

func (m *MmapEntry) mmap(readonly bool) error {
	protection := syscall.PROT_READ
	if !readonly {
		protection |= syscall.PROT_WRITE
	}
	data, err := syscall.Mmap(m.FD, 0, int(m.Size), protection, syscall.MAP_SHARED)
	if err == nil {
		if readonly {
			m.ro = data
		} else {
			m.rw = data
		}
	}
	return err
}

// Connect to IPCClient steps as follows
//
// 1. using unix socket connect to vineyard server
// 2. sending register request to server and get response from server
//
// Note: you should send message's length first to server, then send message
func NewIPCClient(socket string) (*IPCClient, error) {
	c := &IPCClient{}
	c.ClientBase = &ClientBase{}
	c.IPCSocket = socket

	c.unixConn = new(net.UnixConn)
	if err := io.ConnectIPCSocketRetry(socket, &c.unixConn); err != nil {
		return nil, err
	}
	c.conn = c.unixConn

	messageOut := common.WriteRegisterRequest(common.VINEYARD_VERSION_STRING)
	if err := c.doWrite(messageOut); err != nil {
		return nil, err
	}
	var reply common.RegisterReply
	if err := c.doReadReply(&reply); err != nil {
		return nil, err
	}

	c.connected = true
	c.RPCEndpoint = reply.RPCEndpoint
	c.InstanceID = reply.InstanceID
	c.serverVersion = reply.Version

	c.mmapTable = make(map[int]MmapEntry)
	return c, nil
}

func (c *IPCClient) CreateBuffer(size uint64) (blob BlobWriter, err error) {
	if size == 0 {
		return EmptyBlobWriter(c), nil
	}

	messageOut := common.WriteCreateBufferRequest(size)
	if err := c.doWrite(messageOut); err != nil {
		return blob, err
	}
	var reply common.CreateBufferReply
	if err := c.doReadReply(&reply); err != nil {
		return blob, err
	}
	payload := reply.Created
	if size != payload.DataSize {
		return blob, errors.New("data size not match")
	}

	pointer, err := c.mmapToClient(payload.StoreFd, int64(payload.MapSize), false, true)
	if err != nil {
		return blob, err
	}
	v := memory.Slice(pointer, payload.DataOffset, payload.DataSize)
	blob.Reset(payload.ID, payload.DataSize, arrow.NewBufferBytes(v), c.InstanceID)
	return blob, nil
}

func (c *IPCClient) BuildBuffer(address []byte, size uint64) (types.ObjectID, error) {
	if size == 0 {
		empty_blob_writer := EmptyBlobWriter(c)
		return empty_blob_writer.Seal(c)
	}
	if address == nil {
		empty_blob_writer := EmptyBlobWriter(c)
		return empty_blob_writer.Seal(c)
	}
	messageOut := common.WriteCreateBufferRequest(size)
	if err := c.doWrite(messageOut); err != nil {
		return uint64(0), err
	}
	var reply common.CreateBufferReply
	if err := c.doReadReply(&reply); err != nil {
		return uint64(0), err
	}
	payload := reply.Created
	if size != payload.DataSize {
		return uint64(0), errors.New("data size not match")
	}
	pointer, err := c.mmapToClient(payload.StoreFd, int64(payload.MapSize), false, true)
	if err != nil {
		return uint64(0), err
	}
	v := memory.Slice(pointer, payload.DataOffset, payload.DataSize)
	copy(v, address)
	buffer := EmptyBlobWriter(c)
	buffer.Reset(payload.ID, payload.DataSize, arrow.NewBufferBytes(v), c.InstanceID)
	//fmt.Println(buffer.Buffer.Buf())
	return buffer.Seal(c)
}

func (c *IPCClient) GetBuffer(id types.ObjectID, unsafe bool) (Blob, error) {
	buffers, err := c.GetBuffers([]types.ObjectID{id}, unsafe)
	if err != nil {
		return Blob{}, err
	}
	return buffers[id], nil
}

func (c *IPCClient) GetBuffers(
	ids []types.ObjectID,
	unsafe_ bool,
) (map[types.ObjectID]Blob, error) {
	if len(ids) == 0 {
		return nil, nil
	}

	messageOut := common.WriteGetBuffersRequest(ids, unsafe_)
	if err := c.doWrite(messageOut); err != nil {
		return nil, err
	}
	var reply common.GetBuffersReply
	if err := c.doReadReply(&reply); err != nil {
		return nil, err
	}
	buffers := make(map[types.ObjectID]Blob)
	for _, payload := range reply.Payloads {
		if payload.DataSize == 0 {
			buffers[payload.ID] = Blob{Object{payload.ID, nil}, 0, nil}
			continue
		}
		pointer, err := c.mmapToClient(payload.StoreFd, int64(payload.MapSize), true, true)
		if err != nil {
			return nil, err
		}
		v := memory.Slice(pointer, payload.DataOffset, payload.DataSize)
		buffers[payload.ID] = Blob{
			Object{payload.ID, nil},
			payload.DataSize,
			arrow.NewBufferBytes(v),
		}
	}
	return buffers, nil
}

func (c *IPCClient) CreateMetaData(metadata *ObjectMeta) (id types.ObjectID, err error) {
	if !c.connected {
		return types.InvalidObjectID(), NOT_CONNECTED_ERR
	}
	metadata.SetInstanceId(c.InstanceID)
	metadata.SetTransient(true)
	if !metadata.HasKey("nbytes") {
		metadata.SetNBytes(0)
	}
	if metadata.InComplete() {
		_ = c.SyncMetaData()
	}

	id, signature, instanceId, err := c.CreateData(metadata.MetaData())
	if err != nil {
		return id, err
	}

	metadata.SetId(id)
	metadata.SetSignature(signature)
	metadata.SetInstanceId(instanceId)
	if metadata.InComplete() {
		if meta, err := c.GetMetaData(id, false); err != nil {
			return types.InvalidObjectID(), err
		} else {
			*metadata = *meta
		}
	}
	return id, nil
}

func (c *IPCClient) GetMetaData(id types.ObjectID, syncRemote bool) (meta *ObjectMeta, err error) {
	if !c.connected {
		return nil, NOT_CONNECTED_ERR
	}
	metadatas, err := c.GetData([]types.ObjectID{id}, syncRemote, false)
	if err != nil {
		return nil, err
	}
	meta = NewObjectMeta()
	meta.Reset()
	meta.SetMetaData(c.ClientBase, metadatas[0])

	bufferset := meta.GetBuffers()
	buffers, err := c.GetBuffers(bufferset.GetBufferIds(), false)
	if err != nil {
		return nil, err
	}
	for id, buffer := range buffers {
		_ = bufferset.EmplaceBuffer(id, buffer.Buffer)
	}
	return meta, nil
}

func (c *IPCClient) GetObject(id types.ObjectID, object IObject) error {
	if !c.connected {
		return NOT_CONNECTED_ERR
	}
	meta, err := c.GetMetaData(id, false)
	if err != nil {
		return err
	}
	return object.Construct(c, meta)
}

func (c *IPCClient) GetClusterInfo() (map[string]any, error) {
	if !c.connected {
		return nil, NOT_CONNECTED_ERR
	}
	return c.GetClusterMeta()
}

func (c *IPCClient) ListMetadata(pattern string, regex bool, limit int) (map[string]map[string]any, error) {
	if !c.connected {
		return nil, NOT_CONNECTED_ERR
	}
	messageOut := common.WriteListDataRequest(pattern, regex, limit)
	if err := c.doWrite(messageOut); err != nil {
		return nil, err
	}
	var reply common.ListDataReply
	if err := c.doReadReply(&reply); err != nil {
		return nil, err
	}

	return reply.Content, nil
}

func (c *IPCClient) mmapToClient(
	fd int,
	mapSize int64,
	readOnly bool,
	realign bool,
) ([]byte, error) {
	if _, ok := c.mmapTable[fd]; !ok {
		file, err := c.unixConn.File()
		if err != nil {
			return nil, errors.Wrap(err, "failed to obtain native socket")
		}
		clientFd, err := memory.RecvFileDescriptor(int(file.Fd()))
		if err != nil {
			return nil, err
		}
		if clientFd <= 0 {
			return nil, errors.New("receive client fd error")
		}
		entry := MmapEntry{FD: clientFd}
		if realign {
			entry.Size = mapSize + 8 /* sizeof(size_t) */
		} else {
			entry.Size = mapSize
		}
		c.mmapTable[fd] = entry
	}
	entry := c.mmapTable[fd]
	if readOnly && entry.ro == nil || !readOnly && entry.rw == nil {
		if err := entry.mmap(readOnly); err != nil {
			return nil, err
		}
	}
	if readOnly {
		return entry.ro, nil
	} else {
		return entry.rw, nil
	}
}
